src/mpid/ch4/shm/xpmem/xpmem_win.c

```diff
@@ -0,0 +1,187 @@
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+    }
+
+    mpi_errno = MPIR_Group_translate_ranks_impl(shm_group_ptr, shm_comm_ptr->local_size,
+                                                shm_ranks, MPIDI_XPMEM_global.node_group_ptr,
+                                                node_ranks);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    mpi_errno = MPIR_Group_free_impl(shm_group_ptr);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_GET_NODE_RANKS);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_mpi_win_create_hook(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_win_t *xpmem_win = NULL;
+    MPIR_Comm *shm_comm_ptr = win->comm_ptr->node_comm;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    int i;
+    size_t total_shm_size = 0;
+    MPIDIG_win_shared_info_t *shared_table = NULL;
+    int *ranks_in_shm_grp = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_MPI_WIN_CREATE_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_MPI_WIN_CREATE_HOOK);
+    MPIR_CHKPMEM_DECL(2);
+    MPIR_CHKLMEM_DECL(1);
+
+    if (!shm_comm_ptr)
+        goto fn_exit;
+
+    xpmem_win = &win->dev.shm.xpmem;
+    xpmem_win->regist_segs = NULL;
+
+    /* Exchange shared memory region information */
+    MPIR_T_PVAR_TIMER_START(RMA, rma_wincreate_allgather);
+
+    /* Create shared table for later RMA access. Note that shared_table is
+     * a CH4 level structure used also for other window types. CH4 always
+     * initializes shared table for win_allocate and win_allocate_shared because
+     * their shm region are ensured by POSIX. The other window types can only
+     * optionally initialize it in shmmod .*/
+    MPIR_CHKPMEM_MALLOC(MPIDIG_WIN(win, shared_table), MPIDIG_win_shared_info_t *,
+                        sizeof(MPIDIG_win_shared_info_t) * shm_comm_ptr->local_size,
+                        mpi_errno, "shared table", MPL_MEM_RMA);
+    shared_table = MPIDIG_WIN(win, shared_table);
+
+    shared_table[shm_comm_ptr->rank].size = win->size;
+    shared_table[shm_comm_ptr->rank].disp_unit = win->disp_unit;
+    shared_table[shm_comm_ptr->rank].shm_base_addr = win->base;
+    mpi_errno = MPIR_Allgather(MPI_IN_PLACE,
+                               0,
+                               MPI_DATATYPE_NULL,
+                               shared_table,
+                               sizeof(MPIDIG_win_shared_info_t), MPI_BYTE, shm_comm_ptr, &errflag);
+    MPIR_T_PVAR_TIMER_END(RMA, rma_wincreate_allgather);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Skip shared_table initialization if all local processes are zero size. */
+    for (i = 0; i < shm_comm_ptr->local_size; i++)
+        total_shm_size += shared_table[i].size;
+    if (total_shm_size == 0) {
+        MPIDIG_WIN(win, shared_table) = NULL;
+        goto fn_fail;
+    }
```

Comment:
This change is weird to me. It is not a failure case but jump to `fn_fail` ? I'd like to keep the `fn_no_shm` label for code readability. You can change the handling code as below to reuse `fn_fail`.
```
  fn_no_shm:
    MPIDIG_WIN(win, shared_table) = NULL;
  fn_fail:
    MPIR_CHKPMEM_REAP();
    goto fn_exit;
```